### ISSUE TITLE
Case insensitive string search cripples performance in MySQL

### ISSUE THREAD

**Issue by Brutus5000** [yahoo/elide#696]
Hi guys,

we finally made the transition to the Elide 4.x series. Unfortunately a lot of our users complained, that queries that took seconds before now take minutes (which makes most of them return a Gateway timeout by Nginx).

My investigation shows, that the reason is the case insensitive commit https://github.com/yahoo/elide/commit/4f00b7653ddb64a4762aea2d0de197ced68e5ec6, which wraps `==` comparisons for string columns in a `lower()` function. In our MySQL setup this leads to full table scan instead of index usage (even on columns with case insensitive collation! Ironically these already worked case insensitive).

To solve this, their should be some sort of configuration, be it in Elide configuration, on table level or on field level.

If we can agree on an approach, I'd try to implement it.

**Comment by DennisMcWherter** [comment]
This is unfortunate. @clayreimann do you remember why we did this? Is there a reason we're lower-casing these at all? In a case sensitive setup, this may be the wrong thing to do anyway.

Off the top of my head, it **seems** like the logical thing to do is to simply remove the `lower(...)` around the column names. I'll keep thinking about it for a little bit but also may @clayreimann remembers if there was an issue or if it was just converted to lower by mistake.

By the way, my tests also confirm your hypothesis:

**Generated query with `lower(...)` column name:**
```
mysql> explain select post0_.id as id1_2_, post0_.content as content2_2_, post0_.owner_id as owner_id3_2_ from Post post0_ where lower(post0_.content) in (lower('cAsEsInEnSiTiVe'));
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+-------------+
| id | select_type | table  | partitions | type | possible_keys | key  | key_len | ref  | rows | filtered | Extra       |
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+-------------+
|  1 | SIMPLE      | post0_ | NULL       | ALL  | NULL          | NULL | NULL    | NULL |   22 |   100.00 | Using where |
+----+-------------+--------+------------+------+---------------+------+---------+------+------+----------+-------------+
1 row in set, 1 warning (0.00 sec)
```

**Updated query without `lower(...)` column name:**
```
mysql> explain select post0_.id as id1_2_, post0_.content as content2_2_, post0_.owner_id as owner_id3_2_ from Post post0_ where post0_.content in (lower('cAsEsInEnSiTiVe'));
+----+-------------+--------+------------+------+---------------+--------------+---------+-------+------+----------+-------+
| id | select_type | table  | partitions | type | possible_keys | key          | key_len | ref   | rows | filtered | Extra |
+----+-------------+--------+------------+------+---------------+--------------+---------+-------+------+----------+-------+
|  1 | SIMPLE      | post0_ | NULL       | ref  | ContentIndex  | ContentIndex | 403     | const |    1 |   100.00 | NULL  |
+----+-------------+--------+------------+------+---------------+--------------+---------+-------+------+----------+-------+
1 row in set, 1 warning (0.00 sec)
```

**Comment by Brutus5000** [comment]
I think this was discussed here: https://github.com/yahoo/elide/issues/387

**Comment by clayreimann** [comment]
Yep, that's the ticket I was going to go look for. I can think of a couple solutions for this
1. We can add a configuration switch to change which fields get the case insensitive behavior
1. We add a new RSQL operator `===` which maps to `Operator.IN` rather than `Operator. IN_INSENSITIVE `

If a new operator will be sufficient (I hope so because it's probably an order of magnitude less work) I'd start by looking at our RSQL code. Adding a new `CompairsonOperator` and including it should fix your issue. 🤞  https://github.com/yahoo/elide/blob/52222ba866899bf45f8124329bb160978feff023/elide-core/src/main/java/com/yahoo/elide/core/filter/dialect/RSQLFilterDialect.java#L48-L73

**Comment by DennisMcWherter** [comment]
I have read the original ticket and still don't fully understand at this point. To be clear, my issue isn't with case insensitive search on _values_, but on the column itself:

https://github.com/yahoo/elide/blob/9948c4c3f278344dc738d7d5c7ff721315790fee/elide-datastore/elide-datastore-hibernate/src/main/java/com/yahoo/elide/core/filter/HQLFilterOperation.java#L66

Namely, the first `lower(%s)` in that formatted string. If you remove that, this query should perform a proper filter rather than a table scan. What I'm failing to understand is why is this necessary? If the user was querying a case-sensitive schema, (say, `(id, Name)`), we wouldn't expect `... where name = 'abc';` to work in the datastore (it would have to be `Name`). If there isn't a good reason to have this wrapping the column _names_ (not values-- that is absolutely needed), I recommended we do away with that.

Alternatively, if it _is_ required, I highly recommend we lower-case it ourselves in-memory before sending this value in. I don't see a compelling use-case for us to be lower-casing the column name as part of the query to begin with.

Note that we also have to be careful with this code as it can introduce a SQL injection point if not handled properly.

Additional comment: I do see 2 tests fail if I remove this code. Namely,
```
Failed tests: 
  FilterIT.testNonRootFilterInfixInsensitive:1118 expected [0] but found [2]
  FilterIT.testNonRootFilterPrefixInsensitive:1089 expected [0] but found [2]
```

I have not yet had a chance to investigate further.

**Comment by bukajsytlos** [comment]
`lower(column_name) = lower(searched_value)` will lower case data in db and also input data resulting in case insensitive search.

**Comment by DennisMcWherter** [comment]
@bukajsytlos thank you, you're absolutely correct! I was running out the door both mornings going through this and for some reason kept reading the left-hand side (`lower(columName)`) as literally lower'ing the column name (i.e. `lower(Column) -> column`) :man_facepalming:; this makes everything much clearer.

In summary, we wouldn't expect a properly collated index here to work because by calling `lower(columnName)` we skip the index by transforming all the data. This also explains why my original investigation worked (my index was collated with case insensitivity...).

I think @Brutus5000's original suggestion makes most sense to me (@clayreimann's option 1). A new operator sounds complicated here because Elide is doing the wrong thing. Not wrong in the sense of _correctness_ but we're always skipping the index for these searches when that is an unnecessary thing to do. In my opinion, whether or not a search is executed with or without an index is not a burden that should be placed on the caller. The service layer should handle this silently. Certainly, fields without a case insensitive collated index would still require the `lower(columnName)` for correctness, but searching non-indexed fields is always expensive.

I am doing some investigation to see if there is a more elegant way to do this, but configuration is certainly an option (and I'm not convinced it's significantly more work to add). The burden then falls on the service developer rather than the caller where I argue is where the responsibility should lie.

**Comment by DennisMcWherter** [comment]
I've spent a little bit of time now experimenting with potential approaches we can take to _detect_ this. It doesn't look like hibernate provides any information to expose index information. **However,** I believe it is _possible_ to detect the existence of a case-insensitive collation.

In any case, I have outlined the approaches below for a more concrete discussion so we can agree on a path forward and fix this promptly.

### Problem Overview
For **all** case _insensitive_ searches, we are first applying the `lower(...)` function to the data in the DB. As a result, we bypass any indices (even if case insensitive already) and perform a full table scan. For any sizeable table, performance degrades quickly.

### Approach 1: Case Insensitive Search with Configuration
#### Summary
Through some configuration (perhaps, most easily, an annotation) we can provide metadata about the column/index collation. Namely, if it is collated to a case insensitive type, we could avoid `lower(columnName)` and efficiently leverage the index. In the case where it i case sensitive, this column could be `lower()`ed for correctness.

#### Advantages
  * The API caller only ever considers API semantics rather than implementation details
  * Leaves control of actual query performance to backend development

#### Disadvantages
  * Manual labor still required in the backend
  * Additional model annotations required for _expected_ behavior
     * I argue that there would rarely be reasons we would want to skip the index and Elide _should_ be capable of detecting appropriate behavior here. We are forcing model authors to be very explicit for something they would likely never want to additionally customize.

### Approach 2: A New Comparison Operator
#### Summary
We could introduce a new comparison operator that always maps to a "case sensitive" approach: a `===` comparison operator. As a result, this operator will perform a case insensitive search when the table collation is case insensitive.

A _variant_ of this approach is to simply use the case sensitive operators today. Since this change will already require an update in client code, we could avoid any backend update and just point the client to the existing operator. This variant has an obvious disadvantage as being _confusing_, however. Moreover, see my point below about our case sensitive operators.

#### Advantages
  * Quickest/easiest implementation
  * Easy testing since will redirect to existing functionality
  * Doesn't require additional performance or maintenance overhead for collation detection

#### Disadvantages
  * Passes burden to API caller/leaks underlying abstractions
  * No clear API semantic; API caller needs to know table collations
    * Consider the case where the table/index is not collated in a case insensitive way. When the API is called, this will behave differently than when called on a case _sensitive_ column.
  * Requires an update of all client code to adopt this fix
  * In the same way, case sensitive search may be broken as well. This may have been an oversight on our part, but this also disregards table collation.

### Approach 3: Detect Column Case-Sensitivity at Runtime
#### Summary
We can detect table case sensitivity at runtime. There are a few variants to this approach:

  1. **Detect at query time.** Each time a query is issued, the column(s) would be checked for collation information. This would have no impact on startup times, but potentially slow queries a bit. A mitigation to this is that we could lazily load this information and store it in a cache. This would incur a loading cost for the first time a column is accessed, but then be free thereafter.
  1. **Detect at start time.** At Elide startup time we can detect case sensitivity for all columns. This would likely slow down the boot time considerably for large models, however, user-level query performance would not be impacted. While we could provide a mechanism (i.e. time expiration, custom endpoint, etc.) to refresh these caches, collations may _change_ infrequently enough that when they do, perhaps a simple service restart is the correct approach.

While there are variants, the general approach to detecting case sensitivity can be similar. While HQL does not support (afaict) any _generic_ mechanisms for querying indices and fetching column collation (this is specific to each underlying DB), we can instead actually run a test query or queries using [HQL supported expressions](https://docs.jboss.org/hibernate/orm/3.5/reference/en-US/html/queryhql.html#queryhql-expressions). The actual performance of this query may be an exercise in DB tuning, but I believe something like this should provide us what we're looking for:

```sql
mysql> explain select isCaseSensitive from (select (select content from Post limit 1) as elem, (case (select count(*) from Post where content = elem AND lower(content) IN (upper(elem))) WHEN 1 THEN false ELSE true END) as isCaseSensitive) as caseSensitiveIdxCheck;
+----+--------------------+------------+------------+--------+---------------+--------------+---------+------+------+----------+--------------------------+
| id | select_type        | table      | partitions | type   | possible_keys | key          | key_len | ref  | rows | filtered | Extra                    |
+----+--------------------+------------+------------+--------+---------------+--------------+---------+------+------+----------+--------------------------+
|  1 | PRIMARY            | <derived2> | NULL       | system | NULL          | NULL         | NULL    | NULL |    1 |   100.00 | NULL                     |
|  2 | DERIVED            | NULL       | NULL       | NULL   | NULL          | NULL         | NULL    | NULL | NULL |     NULL | No tables used           |
|  4 | DEPENDENT SUBQUERY | Post       | NULL       | index  | ContentIndex  | ContentIndex | 403     | NULL |   22 |   100.00 | Using where; Using index |
|  3 | SUBQUERY           | Post       | NULL       | index  | NULL          | ContentIndex | 403     | NULL |   22 |   100.00 | Using index              |
+----+--------------------+------------+------------+--------+---------------+--------------+---------+------+------+----------+--------------------------+
```

A query such as this scans on the index. Of course, if I remove the index then this query results in scans (as expected):

```
mysql> explain select isCaseSensitive from (select (select content from Post limit 1) as elem, (case (select count(*) from Post where content = elem AND lower(content) IN (upper(elem))) WHEN 1 THEN false ELSE true END) as isCaseSensitive) as caseSensitiveIdxCheck;
+----+--------------------+------------+------------+--------+---------------+------+---------+------+------+----------+----------------+
| id | select_type        | table      | partitions | type   | possible_keys | key  | key_len | ref  | rows | filtered | Extra          |
+----+--------------------+------------+------------+--------+---------------+------+---------+------+------+----------+----------------+
|  1 | PRIMARY            | <derived2> | NULL       | system | NULL          | NULL | NULL    | NULL |    1 |   100.00 | NULL           |
|  2 | DERIVED            | NULL       | NULL       | NULL   | NULL          | NULL | NULL    | NULL | NULL |     NULL | No tables used |
|  4 | DEPENDENT SUBQUERY | Post       | NULL       | ALL    | NULL          | NULL | NULL    | NULL |   22 |   100.00 | Using where    |
|  3 | SUBQUERY           | Post       | NULL       | ALL    | NULL          | NULL | NULL    | NULL |   22 |   100.00 | NULL           |
+----+--------------------+------------+------------+--------+---------------+------+---------+------+------+----------+----------------+
4 rows in set, 3 warnings (0.00 sec)
```

These non-indexed scans can be expensive.

That said, I propose this kind of _testing_ approach since it should answer our actual question in a generic way. Inspecting indices and metadata is very specific to each DB and is a meta-question to what we're actually trying to discover. To avoid this on each startup/query, we could persist this data somewhere (either through a special table in the datastore or on each local disk, etc.). We could then provide either a boot option or some management interface to determine _when_ to re-scan which tables for updated indices. Again, since this information likely doesn't change often, it's probably not strictly necessary to do this on _every_ startup or query.

#### Advantages
  * No code updates required for correct behavior (either in models or clients)
  * More developer-friendly approach (i.e. ease of use)
    * Neither backend nor client-side developers need to tell Elide how to use an index for search
  * Mitigates many synchronization issues (i.e. restart service rather than code fix)
    * i.e. typos, misconfigurations, changing collations, etc.
  * Solution portable without code changes across datastores (i.e. if migrating from MySQL to PostgresSQL).
    * From what I've found, [PostgresSQL has no case insensitive collation today](https://dba.stackexchange.com/questions/156907/equivalent-of-utf8-unicode-ci-collation-in-postgresql/156974#156974).

#### Disadvantages
  * Most complicated approach/more work than either options 1 or 2
  * Will require more extensive testing to guarantee correctness and avoid regression
  * Introduces a performance overhead in finding this information
    * Could be reduced to a one-time cost with some persistence. Additionally could be done as an OOB process if desired.

A little more on the point I mentioned above. Our _case sensitive_ matching may not always be correct either because it also ignores table/column collation (I haven't actually tested this yet, but this is my hypothesis). As a result, I do not believe option 2 is truly viable. Our API semantics could be inverted based on table collation.

I advocate for approach 3. It is certainly more work, however, may ensure we preserve defined semantics for the API.

Does anyone have any additional thoughts on this? @aklish @wcekan do either of you see any major issues with approach 3 to prefer approach 1 (or have an alternative approach)? I know this post is a bit lengthy, but I wanted to give us enough information to discuss to choose a path forward. Thanks!

**Comment by bukajsytlos** [comment]
I would leave default compare without any special treat.
To make case insensitive search three approaches come to my mind:

1. CI collation - can only perform CI search
2. functional index + lower(column) - can perform both CI and CS search
3. generated column + index - can perform both CI and CS search

In first approach I would not expect anything from elide.
In second approach I would be glad if elide could generate correct query.
In third approach I would need elide to use different column for case insensitive search.

Next question is if this should be decided "compile-time" (config/annotation/magic) or run-time (operator), therefore by client. I don't think there is a need for client to be able to decide per use-case, but could be nice to have.

PS: I am not sure I understand your third approach. If Elide detects it is effective to execute CI search it would prefer this without option to do CS search?

**Comment by DennisMcWherter** [comment]
> CI collation - can only perform CI search

We could impose this constraint. The implementation then falls back as I recommended earlier to simply remove the `lower(...)` calls.

For approaches 2 and 3 you've suggested, I think the selection of database matters. iirc, [MySQL (5.7.5+) now supports generated/virtual columns](https://dev.mysql.com/doc/refman/5.7/en/create-table-generated-columns.html), but not functional indices. The [inverse is true for PostgresSQL](https://www.postgresql.org/docs/9.1/static/indexes-expressional.html). Since this store _intends_ to support all hibernate-supported databases, we'll probably want to avoid/fix existing assumptions about the actual underlying database.

> PS: I am not sure I understand your third approach. If Elide detects it is effective to execute CI search it would prefer this without option to do CS search?

This is a good call out; I hadn't thought through the CS case fully. Since Elide doesn't actually perform this filtering in memory for all cases (in the back of my head I was thinking it was), perhaps this approach still won't work well. Even so, in-memory filtering is insufficient for correctness when paired with pagination.

In the case where one wants to perform a _case sensitive search_ with a CI collation, we would still be broken today. In light of this, approach 3 would not be sufficient on its own.

What concerns me more is that we have no way of specifying a case **sensitive** search on a string today through [RSQL as far as I can tell](https://github.com/yahoo/elide/blob/master/elide-core/src/main/java/com/yahoo/elide/core/filter/dialect/RSQLFilterDialect.java#L308) (`isStringLike` appears to be true for all strings).

Perhaps a hybrid approach actually is best here due to the variety of different features in different databases. Allowing the model author to, perhaps, specify how to query the table in the case-sensitive or insensitive situations would be the most effective option. The author could then model indices, generated columns, etc. as required in the database and merely instruct Elide how to query more efficiently.

**Comment by Brutus5000** [comment]
I just thought about the implications of a second operator on database level. I see two cases:

a) Column has case sensitive collation:
a.1) User requests case sensitive filter -> all works out
a.2) User requests case insensitive filter -> we need perform lower function on the column which will cause heavy load and long query times on big tables (at least in MySQL)

b) Column has case insensitive collation:
b.1) User requests case sensitive filter -> this can't be done on database layer. In memory processing required which brings problem with pagination as described by @DennisMcWherter 
b.2) User requests case insensitive filter -> all works out

As far as I can see adding a new operator brings a lot of pitfalls to the consumer of the api. Further more there is no way to describe in the swagger ui which operator the consumer should choose for maximum performance.

I would suggest to keep the `==` as the default operator using the collation of the database. The database designer should be able to consider the pros and cons of the collation. If the consumer wants to enfore CI or CS search he could use a regex operator.

**Comment by aklish** [comment]
[FIQL](https://tools.ietf.org/html/draft-nottingham-atompub-fiql-00) defines `==`as case insensitive:

` Simple text comparisons allow for case insensitive, non-language-
   specific searching of the textual content of selected nodes.`

I don't think we can redefine `==` to be anything else.

Ideally, I wish there was a convenient way for Elide to know the collation spec of a given column - and then determine if `lower` was necessary or not.  

When `lower` causes performance problems, I believe the DB can create a 2nd functional index or generated index to solve the performance issue.  The primary thing that is missing is Elide being smart and not putting `lower` on CI columns (where it is not needed).

I couldn't find anything obvious in JPA to indicate collation other than maybe the `@Column` annotation.  That could be parsed maybe to identify whether the column is CI or not.   However, I doubt this parsing logic will be portable (since it is native SQL) - so not ideal.

Shy of that, we could introduce a new annotation to specify if a column is case sensitive or not.  Querying the DB seems overkill to me.  

Another simple option would be for Elide to have a setting which describes the default case sensitivity of the DB.  It could then include/exclude `lower` based on that setting alone (without being column aware).  My guess is that would work in the common case.

My preferences in order:
1. Elide global setting for default DB case sensitivity (simple but does not account for all schemas)
2. New Elide Annotation for the Hibernate/JPA stores only.
3. New RSQL === operator that is case sensitive.  If the column is already case insensitive, this won't work. (I'm guessing somebody will file an issue that this doesn't work down the road).
4. Elide parses and understands `@Column` column description text (I'm worried this will not be portable and cause us headaches down the road).  

Thoughts?

**Comment by bukajsytlos** [comment]
I still don't think it's responsibility of Elide. Let Elide users decide if and how they want to achieve CI (FIQL compliance).
If I put CS collation, I don't want Elide to be "smart" and put lower there. However Elide should provide some sort of mechanism (lower) to enable CI on CS columns.

**Comment by clayreimann** [comment]
Fixed by #706
